### PR TITLE
added RemoveUnusedEnzymeOpsPass

### PR DESCRIPTION
**Pull request by umatin**
This adds a pass which deletes gradient ops which are unnecessary.

### CODE REVIEW

File: enzyme/Enzyme/MLIR/Passes/RemoveUnusedEnzymeOps.cpp [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+#include "llvm/Support/raw_ostream.h"
+#include "mlir/IR/Dominance.h"
+
+using namespace mlir;
+using namespace enzyme;
+using llvm::errs;
+namespace {
+
+// TODO: Expand to region branches??
+bool reachable(Operation * a, Operation * b){
```
Comment (pengmai) [inline enzyme/Enzyme/MLIR/Passes/RemoveUnusedEnzymeOps.cpp]: Is this function meaningfully different than `DominanceInfo`?

File: enzyme/Enzyme/MLIR/Passes/RemoveUnusedEnzymeOps.cpp [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+          if (auto type = dyn_cast<enzyme::GradientType>(initOp.getType())) {
+            Value v = initOp;
+            bool replaceable = true;
+            for(Operation * userSet : v.getUsers()){
+              if (auto setOp = dyn_cast<enzyme::SetOp>(userSet)) {
+                for(Operation * userGet : v.getUsers()){
+                  if (auto getOp = dyn_cast<enzyme::GetOp>(userGet)) {
+                    bool propertyA = dInfo.dominates(userSet, userGet) && !reachable(getOp, setOp);
+                    bool propertyB = !reachable(setOp, getOp);
+                    if (!propertyA && !propertyB){
```
Comment (pengmai) [inline enzyme/Enzyme/MLIR/Passes/RemoveUnusedEnzymeOps.cpp]: I'm finding this logic challenging to parse (`if not (userSet dominates userGet and getOp does not reach setOp))...`?). Perhaps some combination of naming, comments, and/or breaking out into functions would clarify this.

File: enzyme/Enzyme/MLIR/Passes/RemoveUnusedEnzymeOps.cpp [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+                  }
+                }
+              }
+            }
+            if(replaceable){
+              // Do replacing
+              for(Operation * userGet : v.getUsers()){
+                if (auto getOp = dyn_cast<enzyme::GetOp>(userGet)) {
+                  Operation * closestSetOp = nullptr;
+                  for(Operation * userSet : v.getUsers()){
```
Comment (pengmai) [inline enzyme/Enzyme/MLIR/Passes/RemoveUnusedEnzymeOps.cpp]: I'd also recommend breaking this "find closest set op" logic out into a function. In addition to clarity, it seems like it'd be reusable in other cases than just handling Gradient types.

File: enzyme/Enzyme/MLIR/Passes/RemoveUnusedEnzymeOps.cpp (hunk 1 above)
Comment (umatin) [reply enzyme/Enzyme/MLIR/Passes/RemoveUnusedEnzymeOps.cpp]: I tried to find something and came up short, so I think so.

File: enzyme/Enzyme/MLIR/Passes/RemoveUnusedEnzymeOps.cpp (hunk 3 above)
Comment (umatin) [reply enzyme/Enzyme/MLIR/Passes/RemoveUnusedEnzymeOps.cpp]: Good idea 👍

File: enzyme/Enzyme/MLIR/Passes/RemoveUnusedEnzymeOps.cpp [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+//------------------ //
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file implements a pass to lower custom ops generated by the Enzyme AD
+// procedure to the MemRef dialect.
```
Comment (pengmai) [inline enzyme/Enzyme/MLIR/Passes/RemoveUnusedEnzymeOps.cpp:12]: Nit: this looks out of sync

**Review by pengmai — APPROVED** [review]

**Pull request merged**